bokeh/io.py

```diff
@@ -516,7 +535,7 @@ def push_notebook(document=None, state=None, handle=None):
         msg = Document._compute_patch_between_json(handle.json, to_json)
     handle.comms.send(json.dumps(msg))
     handle._doc = document
-    handle._json = to_json
+    handle._json[id(document)] = to_json
```

Comment:
That branch is to handle the case where the document itself changes (e.g. someone creates `Document()` by hand, I guess). But that is not this case. In this case, there is only `curdoc` and it has output that might span several output cells. But the comms handler for a given cell only knows about what it originally showed. We need to diff against the "last shown json" so that we don't sent updates for models that are shown in other cells by other comms handlers (which is what was happening and breaking things). `handle._json is basically just a cache of the current document serialized, it does't seem especially any worse that having the document itself kept around. It's like like a completely unbounded accumulation of "events"